src/services/download/mod.rs

```diff
@@ -204,6 +201,25 @@ fn download_future_stream(
     Box::new(ret)
 }
 
+async fn download_stream(
+    source: SourceFileId,
+    stream: impl futures::stream::Stream<Item = Result<bytes::Bytes, DownloadError>>,
+    destination: PathBuf,
+) -> Result<DownloadStatus, DownloadError> {
+    // All file I/O in this function is blocking!
+    log::trace!("Downloading from {}", source);
+    let mut file =
+        std::fs::File::create(&destination).context(DownloadErrorKind::BadDestination)?;
+    let mut stream = Box::pin(stream);
```

Comment:
`StreamExt::next` needs the stream to be `Unpin`.  So I pinned it.

I think as we discovered before, when a struct is `Unpin` that doesn't mean the fields are as well.  Probably that's why `stream` here is not `Unpin` automatically?

Reply:
Rather than using `impl Stream`, I tried with a generic function and then you don't need the Box::pin, since the `Unpin` requirement then simply propagates up. With imports (which I'd prefer) and a little bit of reformatting, this would then read:

```rust
use bytes::Bytes;
use futures::Stream;
use std::fs::File;

async fn download_stream<S>(
    source: SourceFileId,
    stream: S,
    destination: PathBuf,
) -> Result<DownloadStatus, DownloadError>
where
    S: Stream<Item = Result<Bytes, DownloadError>>,
{
    // All file I/O in this function is blocking!
    log::trace!("Downloading from {}", source);
    let mut file = File::create(&destination).context(DownloadErrorKind::BadDestination)?;
    let mut stream = Box::pin(stream);

    while let Some(chunk) = stream.next().await {
        file.write_all(chunk?.as_ref())
            .context(DownloadErrorKind::Write)?;
    }

    Ok(DownloadStatus::Completed)
}
```

Reply:
you still have the `Box::pin` in there though ;-)

Reply:
:/ Silly me, updated the example.

Reply:
Alright, in this case, using [`futures::pin_mut!`](https://docs.rs/futures/0.3.5/futures/macro.pin_mut.html) is the best option. Essentially the same as `Box::pin`, just without the box.